File: android/app/src/main/java/com/bensbible/app/viewmodel/SearchViewModel.kt
package com.bensbible.app.viewmodel

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bensbible.app.data.BibleDataService
import com.bensbible.app.model.BookGroup
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

enum class SearchMode { PHRASE, ALL_WORDS }

data class SearchResult(
    val bookName: String,
    val chapter: Int,
    val verse: Int,
    val text: String
) {
    val reference: String get() = "$bookName $chapter:$verse"
}

class SearchViewModel(
    private val dataService: BibleDataService
) : ViewModel() {

    var query by mutableStateOf("")
        private set

    var results by mutableStateOf<List<SearchResult>>(emptyList())
        private set

    var isSearching by mutableStateOf(false)
        private set

    var selectedGroup by mutableStateOf(BookGroup.ALL)
        private set

    var searchMode by mutableStateOf(SearchMode.PHRASE)
        private set

    private var searchJob: Job? = null

    fun onSearchModeChange(mode: SearchMode) {
        searchMode = mode
        searchJob?.cancel()
        val trimmed = query.trim()
        if (trimmed.isEmpty()) return
        isSearching = true
        searchJob = viewModelScope.launch {
            performSearch(trimmed)
        }
    }

    fun onGroupChange(group: BookGroup) {
        selectedGroup = group
        searchJob?.cancel()
        val trimmed = query.trim()
        if (trimmed.isEmpty()) return
        isSearching = true
        searchJob = viewModelScope.launch {
            performSearch(trimmed)
        }
    }

    fun onQueryChange(newQuery: String) {
        query = newQuery
        searchJob?.cancel()

        val trimmed = newQuery.trim()
        if (trimmed.isEmpty()) {
            results = emptyList()
            isSearching = false
            return
        }

        isSearching = true
        searchJob = viewModelScope.launch {
            delay(300)
            performSearch(trimmed)
        }
    }

    private suspend fun performSearch(query: String) {
        val matches = withContext(Dispatchers.Default) {
            val allBooks = dataService.loadBookNames()
            val bookNames = selectedGroup.filterBooks(allBooks)
            val found = mutableListOf<SearchResult>()
            for (bookName in bookNames) {
                val book = dataService.loadBook(bookName)
                for (chapter in book.chapters) {
                    for (verse in chapter.verses) {
                        val matched = when (searchMode) {
                            SearchMode.PHRASE -> verse.text.contains(query, ignoreCase = true)
                            SearchMode.ALL_WORDS -> query.trim().split("\\s+".toRegex())
                                .all { word -> verse.text.contains(word, ignoreCase = true) }
                        }
                        if (matched) {
                            found.add(
                                SearchResult(
                                    bookName = bookName,
                                    chapter = chapter.number,
                                    verse = verse.number,
                                    text = verse.text
                                )
                            )
                        }
                    }
                }
            }
            found
        }
        results = matches
        isSearching = false
    }
}

File: BensBible2/ViewModels/SearchViewModel.swift
import Foundation

enum SearchMode: String, CaseIterable {
    case phrase = "Phrase"
    case allWords = "All Words"
}

struct SearchResult: Identifiable {
    let id = UUID()
    let bookName: String
    let chapter: Int
    let verse: Int
    let text: String

    var reference: String { "\(bookName) \(chapter):\(verse)" }
}

@Observable
final class SearchViewModel {
    private let dataService: BibleDataService
    private var searchTask: Task<Void, Never>?

    var query: String = "" {
        didSet { debounceSearch() }
    }
    var results: [SearchResult] = []
    var navigationTarget: BibleLocation? = nil
    var isSearching: Bool = false
    var selectedGroup: BookGroup = .all
    var searchMode: SearchMode = .phrase {
        didSet { triggerSearch() }
    }

    init(dataService: BibleDataService = LocalBibleDataService()) {
        self.dataService = dataService
    }

    func selectGroup(_ group: BookGroup) {
        selectedGroup = group
        searchTask?.cancel()
        let currentQuery = query.trimmingCharacters(in: .whitespaces)
        guard !currentQuery.isEmpty else { return }
        isSearching = true
        searchTask = Task {
            await performSearch(query: currentQuery)
        }
    }

    private func triggerSearch() {
        searchTask?.cancel()
        let currentQuery = query.trimmingCharacters(in: .whitespaces)
        guard !currentQuery.isEmpty else { return }
        isSearching = true
        searchTask = Task {
            await performSearch(query: currentQuery)
        }
    }

    private func debounceSearch() {
        searchTask?.cancel()
        let currentQuery = query.trimmingCharacters(in: .whitespaces)

        guard !currentQuery.isEmpty else {
            results = []
            navigationTarget = nil
            isSearching = false
            return
        }

        navigationTarget = nil
        isSearching = true
        searchTask = Task {
            try? await Task.sleep(nanoseconds: 300_000_000)
            guard !Task.isCancelled else { return }
            await performSearch(query: currentQuery)
        }
    }

    private func performSearch(query: String) async {
        do {
            let allBooks = try dataService.loadBookNames()
            let navTarget = parseNavigationTarget(query: query, allBooks: allBooks)
            let bookNames = selectedGroup.filterBooks(from: allBooks)
            var matches: [SearchResult] = []

            for bookName in bookNames {
                if Task.isCancelled { return }
                let book = try dataService.loadBook(named: bookName)
                for chapter in book.chapters {
                    for verse in chapter.verses {
                        let matched: Bool
                        switch searchMode {
                        case .phrase:
                            matched = verse.text.localizedCaseInsensitiveContains(query)
                        case .allWords:
                            let words = query.split(separator: " ").map(String.init)
                            matched = words.allSatisfy { verse.text.localizedCaseInsensitiveContains($0) }
                        }
                        if matched {
                            matches.append(SearchResult(
                                bookName: bookName,
                                chapter: chapter.number,
                                verse: verse.number,
                                text: verse.text
                            ))
                        }
                    }
                }
            }

            let finalMatches = matches
            guard !Task.isCancelled else { return }
            await MainActor.run {
                self.results = finalMatches
                self.navigationTarget = navTarget
                self.isSearching = false
            }
        } catch {
            await MainActor.run {
                self.results = []
                self.navigationTarget = nil
                self.isSearching = false
            }
        }
    }

    private func parseNavigationTarget(query: String, allBooks: [String]) -> BibleLocation? {
        let parts = query.trimmingCharacters(in: .whitespaces)
            .components(separatedBy: .whitespaces).filter { !$0.isEmpty }
        guard parts.count >= 2 else { return nil }

        let lastPart = parts.last!
        let cv = lastPart.components(separatedBy: ":")
        guard let chapter = Int(cv[0]), chapter > 0 else { return nil }
        let verse = cv.count > 1 ? Int(cv[1]) : nil

        let potentialBook = parts.dropLast().joined(separator: " ")
        let aliases = ["Psalm": "Psalms"]
        let resolvedBook = aliases[potentialBook] ?? potentialBook

        guard let book = allBooks.first(where: {
            $0.localizedCaseInsensitiveCompare(resolvedBook) == .orderedSame
        }) else { return nil }

        return BibleLocation(bookName: book, chapterNumber: chapter, verseNumber: verse)
    }
}
